Implement the new selection algorithm

cescoffier opened the pull request:
Related to #2737.

cescoffier committed:
Implementation of the new extension selection algorithm as proposed in #2737.

Impact of the new selection algorithm on the guides.

Most of it is purely cosmetic and allow using shorter names. Some tweaks are required to avoid ambiguity.

Do not use dash in bean validation.

Remove the swagger-ui short name. Swagger may mean something else.

gsmet commented on docs/src/main/asciidoc/using-vertx.adoc:
```diff
@@ -19,15 +19,15 @@ If you are creating a new project, set the `extensions` parameter are follows:
     -DprojectGroupId=org.acme \
     -DprojectArtifactId=vertx-quickstart \
-    -Dextensions="vertx"
+    -Dextensions="vert.x"
```
Wondering if it's a good idea to have the dot here? I suppose not having the dot would work anyway as you have an exact match for `quarkus-vertx`?

cescoffier replied:
The real/official name includes the dot. That's why. @vietj WDYT?

cescoffier replied:
Just removed the short name as it's not required anymore.

This case is handled by the partial name match.

gsmet reviewed (commented):
Ah, you beat me to it, thanks!

I added two minor comments. It looks good otherwise but I'll let @emmanuelbernard do the review.

mkouba commented:
Hm, I wonder if we should display the short name and possibly also tags in `list-extensions` (fulll format)?

emmanuelbernard commented on devtools/common/src/main/filtered/extensions.json:
```diff
@@ -105,6 +107,7 @@
   },
   {
     "name": "Hibernate Validator",
+    "shortName": "bean-validation",
```
I don't think people write bean-validation. It's `Bean Validation`

cescoffier replied:
Fixed

emmanuelbernard commented on devtools/common/src/main/filtered/extensions.json:
```diff
@@ -432,6 +437,7 @@
   },
   {
     "name": "Swagger UI",
+    "shortName": "swagger",
```
Do we really want the Swagger UI component to trump all other "swagger" stuff? Isn't swagger more than just a UI component?

cescoffier replied:
I don't really know. We always refer to _swagger_. Might be better to not use it there.

kenfinnigan replied:
The other bit that is `swagger` from the community we have in the form of `openapi`, so I don't think we'd add a pure `swagger` extension. Though it may be confusing to some to not have UI in there

cescoffier replied:
Yes, I've removed the short name.

emmanuelbernard commented on devtools/common/src/main/java/io/quarkus/cli/commands/AddExtensions.java:
```diff
@@ -31,57 +33,96 @@ public AddExtensions(final ProjectWriter writer, final String pom) throws IOExce
+        List<Extension> matchesLabels = extensions.stream()
+                .filter(extension -> extension.labels().contains(q)).collect(Collectors.toList());
+
+        Set<Extension> candidates = new LinkedHashSet<>();
```
The old algorithm had a match on partial artifactId and partial name match. It seems to be lost in this new approach.

e.g. `orm-pana` => `hibernate-orm-panache`, `jdbc-postgre` -> JDBC PostgreSQL driver

cescoffier replied:
I've re-introduced this behavior.

cescoffier replied:
(with tests of course).

emmanuelbernard commented on devtools/common/src/main/java/io/quarkus/cli/commands/AddExtensions.java:
```diff
@@ -31,57 +33,96 @@ public AddExtensions(final ProjectWriter writer, final String pom) throws IOExce
+    private static boolean matchesArtifactId(String artifactId, String q) {
+        return (artifactId.equalsIgnoreCase(q) ||
+                artifactId.equalsIgnoreCase("quarkus-" + q) ||
+                artifactId.equalsIgnoreCase("quarkus-smallrye-" + q));
```
Is there consensus on that one. I through the idea but was not 100% sure myself.

emmanuelbernard replied:
For example the `rest-client`, we are discuss which REST Client we really want. What if we don't go for the smallrye one for the selected one?

cescoffier replied:
That would mean we would have 2 rest client? 

Where this is useful is to include things like `metrics` or `health` using these simple words. Now we can have the same behavior using `shortName` for these. 

@kenfinnigan WDYT?

kenfinnigan replied:
Currently we have the SmallRye REST Client, once there's a RESTEasy 4.x release with the MP REST Client impl there we would remove SmallRye REST Client and only go with RESTEasy one

cescoffier replied:
So do we keep this match or not?

If we drop it, we may want to add short names to the different smallrye implementation (metrics, health...).

cescoffier replied:
(Forgot to say, no strong opinion)

kenfinnigan replied:
I can see it makes sense, though it saddens me that SmallRye becomes hidden completely in Quarkus.

cescoffier replied:
The groupIds of the dependency are still visible (that does not change). It's only how the user selects the extension.

1. Smallrye is still visible in list-extension.
2. The user can do `add-extension -Dextension=smallrye-health` (he can even use the full artifactId)
3. The only difference is: the user can also do: `add-extension -Dextension=health` which would print on screen: smallrye-health has been added (or something like this, but the extension name is used in this case).

emmanuelbernard replied:
Right my point is that if at some point we don’t want the SR version of something then we would be stuck. I prefer that we use short names then for the ones we want to shorten and recommend. Instead of that rule that is.

cescoffier replied:
Ok, will update.

cescoffier replied:
Removed and short names added.

emmanuelbernard commented on devtools/common/src/main/java/io/quarkus/cli/commands/AddExtensions.java:
```diff
@@ -31,57 +33,96 @@ public AddExtensions(final ProjectWriter writer, final String pom) throws IOExce
+                // We have 3 cases, we can still have a single candidate, but the match is on label
+                // or we have several candidates, or none
+                Set<Extension> candidates = result.getExtensions();
+                if (candidates.isEmpty() && query.contains(":")) {
```
`query.contains(":")` always beats candidates I think. It's an absolute name.

cescoffier replied:
Agreed.

cescoffier replied:
in addition, the insertion must avoid duplicates, so 2 bugs.

cescoffier replied:
Fixed

emmanuelbernard commented on devtools/common/src/main/java/io/quarkus/cli/commands/AddExtensions.java:
```diff
@@ -31,57 +33,96 @@ public AddExtensions(final ProjectWriter writer, final String pom) throws IOExce
-
-            if (matches.size() > 1) {
+        for (String query : extensions) {
+            List<Extension> registry = MojoUtils.loadExtensions();
```
Are you loading extensions every single loop because an previous extension might have been added?

cescoffier replied:
Fixed.

emmanuelbernard commented on devtools/common/src/main/java/io/quarkus/cli/commands/AddExtensions.java:
```diff
@@ -31,57 +33,96 @@ public AddExtensions(final ProjectWriter writer, final String pom) throws IOExce
+                    print(NOK + " Cannot find a dependency matching '" + query + "', maybe a typo?");
+                    success = false;
+                } else if (candidates.size() == 1) {
+                    sb.append(NOK).append(" One extension matching '").append(query).append("'");
```
I don't understand if there is a single matching element, why do you NOK it?

cescoffier replied:
because it's typically a single match on labels. (you wrote the proposal ;-))

cescoffier replied:
Should we change this? 

Imagine that there is a single match but it comes from labels. Should we ask the user to be more explicit?

emmanuelbernard replied:
Maybe I don’t understand when this case is happening then. 
I think in the proposal I proposed to not search on labels at all. Here you search but complain to the user? Or is it something else ?

cescoffier replied:
You are right it still looks for labels and _indicate_ to the user the matching extensions. 
I will remove the label lookup.

cescoffier replied:
Removed.

emmanuelbernard replied:
Hey @cescoffier I've been thinking about it some more.
I now (aka today ;) ) that removing labels might be a bit aggressive. For example if one wants to find smallrye-openapi when one types Swagger, the  way to do it is to have it in the name. Which is constraining. So Maybe a label search make sense afterall.
Since they are treated less importantly than shortname and "full" name match, this might be acceptable.

emmanuelbernard replied:
The one option we have is to auto accept them, or do like you did  here and stop the algorithm by asking to select/confirm the extension explicitly.
I was fine with option 1 but I can be convinced otherwise. If we go with option 2, we need a clearer text to describe what is going on as the general non matching text is confusing.

emmanuelbernard commented on devtools/common/src/main/java/io/quarkus/cli/commands/AddExtensions.java:
```diff
@@ -91,7 +132,32 @@ public boolean addExtensions(final Set<String> extensions) throws IOException {
+                            isDefinedInBom(dependenciesFromBom, extension)));
+            return true;
+        } else {
+            // Don't set success to false as it's should be idempotent.
```
you say don't set it to false but you do set it to false. I'm missing something.

cescoffier replied:
wrong place for this comment.

cescoffier replied:
Fixed.

emmanuelbernard commented on devtools/common/src/test/java/io/quarkus/cli/commands/AddExtensionsTest.java:
```diff
@@ -4,20 +4,20 @@
 import java.io.IOException;
-import java.util.HashMap;
-import java.util.HashSet;
+import java.util.*;
```
Do we do .* ?

cescoffier replied:
NO!

cescoffier replied:
Fixed

emmanuelbernard commented on devtools/common/src/test/java/io/quarkus/cli/commands/AddExtensionsTest.java:
```diff
@@ -29,17 +29,236 @@ public void addExtension() throws IOException {
         File pomFile = new File(pom.getAbsolutePath());
         new AddExtensions(new FileProjectWriter(pomFile.getParentFile()), pomFile.getName())
-                .addExtensions(new HashSet<>(asList("jdbc-postgre", "agroal", "quarkus-arc", " hibernate-validator")));
+                .addExtensions(new HashSet<>(asList("jdbc-postgre", "agroal", "quarkus-arc", " hibernate-validator",
```
+1

cescoffier replied:
That's a great point. I've added test to reproduce the conflicts we had, but of course / unfortunately, I'm not covering all of them. I'm going to add it.

cescoffier replied:
Done, I've added a specific test.

Now, `arc` add `quarkus-arc`. To get `hibernate-search-elasticsearch`, you need to have the `elasticsearch` query (partial match on the artifactId or name).

emmanuelbernard commented:
> Hm, I wonder if we should display the short name and possibly also tags in `list-extensions` (fulll format)?

list-extensions is already too long. I'd use shortname and labels in the web ui as tags and as search restrictions. Not necessarily expose them.

cescoffier committed:
Support partial matches in the name, artifactId or shortName

If a query is a GAV, insert the dependency as a GAV. Also, avoid reloading the extension list every time.

When the query is containg in the artifact id or in the name of the extension.

cescoffier commented:
If you find conflicts or odd behavior in the currently used algorithm (in 0.16.1), let me know we can add a test and check.

cescoffier committed:
Add short names for smallrye dependencies in the extension list
Disable the label match by default

gsmet commented:
@cescoffier could you squash things a bit. Typically the `Apply formater` commit should go away. As for the others, I let you judge if they have values. I think I would squash everything to have one commit with all the tuning but feel free to do otherwise.

cescoffier commented:
@gsmet I would squash this into a single commit no? The different iterations are not very meaningful. (Github can to it).

emmanuelbernard commented:
I looked  fast but I did not see implemented the approach where the short name was used as the default value for a given choice menu.
In https://github.com/quarkusio/quarkus/issues/2737#issuecomment-499598004 it's described as option 2.

Were do we stand on option 1 vs option 2 collectively. We we discussed it by voice, I think the compromise was for option 2 as option 1 was denying the notion of a kafka label for searches.

cescoffier commented:
Option 2 requires a prompt and changing completely how the Maven plugin works. That's why I picked 1 for now until we find the right approach to implement 2.

gsmet commented:
Moving to 0.18.0. I discussed with @cescoffier and the situation is unclear. I think we need a call between @emmanuelbernard  and @cescoffier  to clarify the situation and converge.

FroMage commented:
Also needs to be rebased now, apparently.

cescoffier commented:
Commit squashed, rebase done.
